File: quickSort.py
# Implementation of the quick sort algorithm
def quickSort(theList, left, right):
	# left = 0
	# right = len(theList) - 1

	if left < right:
		pi = partition(theList, left, right)

		quickSort(theList, left, pi-1)
		quickSort(theList, pi+1, right)

	return theList


def partition(arr, left, right):
	i = left - 1
	pivot = arr[right]

	for j in range(left, right):
		if arr[j] > pivot: # > descending, < ascending
			i += 1
			arr[i], arr[j] = arr[j], arr[i]

	arr[i+1], arr[right] = arr[right], arr[i+1]
	return i + 1

def kLargestNumber(arr, k):
	left = 0
	right = len(arr) - 1

	while True:
		pi = partition(arr, left, right)
		if pi > k-1:
			right = pi - 1
		elif pi < k-1:
			left = pi + 1
		else:
			break

	return arr[pi]

if __name__ == "__main__":

	seq = [1,5,8,2,50,3937,9,0,83,8]
	print(quickSort(seq, 0, len(seq)-1))

	seq = [1,5,8,2,50,3937,9,0,83,8]
	assert kLargestNumber(seq, 2) == 83, "Klargest fail"
	print("DONE")

	

File: quickSort.cpp
#include <iostream>
#include <vector>

using namespace std;

void swap(int* a, int* b ) {
	int t = *(a);
	*(a) = *(b);
	*(b) = t;
};

int partition(vector<int>& arr, int left, int right) {
	int i = left - 1;
	int pivot = arr[right];

	for (int j = left; j < right; ++j) {
		if (arr[j] < pivot) { // < ascending, > descending
			i++;
			swap( &arr[i], &arr[j] );
		}
	}

	swap( &arr[i+1], &arr[right] );
	return i+1;
};

void quickSort(vector<int>& arr, int left, int right) {
	if ( left >= right ) return;

	int pi = partition( arr, left, right );
	quickSort( arr, left, pi-1 );
	quickSort( arr, pi+1, right );

	return;
};

int main(int argc, char const *argv[])
{
	
	vector<int> vec { 2,43,6,5,7,6,9,4,5,77,3,43,5,2,3 };

	quickSort(vec, 0, vec.size()-1);

	for (int n : vec) cout << n << " ";

	cout << endl;

	return 0;
}


